File: BigO/Python/add.py
import time

# Number of operations is (eventually) bounded by a multiple of n (like 10n)
# O(n)


def addUpTo(n):
    total = 0
    for i in range(n + 1):
        total += i
    return total

# Always 3 operators
# O(1)


def addUpTo2(n):
    return (n * (n + 1)) / 2


t1 = time.time()
addUpTo(1000000000)
t2 = time.time()
print(f'Time elapsed: {t2 - t1} seconds')  # 57-58 seconds

t3 = time.time()
addUpTo2(1000000000)
t4 = time.time()
print(f'Time elapsed: {t4 - t3} seconds')  # 1.19 seconds

# def recurse(list, target):
#     # base case is empty list
#     if len(list) == 0:
#         return False
#     elif list[0] == target:
#         return True
#     else:
#         return recurse(list[1:], target)

# my_list = [2, 4, 5, 3]

File: BigO/JavaScript/add.js
// Number of operations is (eventually) bounded by a multiple of n (like 10n)
// O(n)
function addUpTo(n) {
  let total = 0;
  for (let i = 1; i <= n; i++) {
    total += i;
  }
  return total;
}
// Always 3 operators
// O(1)
function addUpTo2(n) {
  return (n * (n + 1)) / 2;
}
// Time elapsed was 2.27 seconds
let t1 = performance.now();
console.log(addUpTo(1000000000));
let t2 = performance.now();
console.log(`Time Elapsed: ${(t2 - t1) / 1000} seconds`);

// Time elapsed was 0.00012
let t3 = performance.now();
console.log(addUpTo2(1000000000));
let t4 = performance.now();
console.log(`Time Elapsed: ${(t4 - t3) / 1000} seconds`);
